File: spring/src/main/java/io/github/wavesync/repository/ResonatorMasterRepository.java
// MIGRATED to fastapi/app/repositories/resonator_master_repository.py
package io.github.wavesync.repository;
import io.github.wavesync.dto.response.ResonatorSummaryResponseDto;
import io.github.wavesync.entity.ResonatorMaster;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;
import java.util.List;


@Repository
public interface ResonatorMasterRepository extends JpaRepository<ResonatorMaster, Long> {

    ResonatorMaster findByName(String name);

    @Query("""
    SELECT new io.github.wavesync.dto.response.ResonatorSummaryResponseDto(
        ur.id,
        m.name,
        m.rarity,
        m.releaseVersion,
        m.thumbnailImage
    )
    FROM ResonatorMaster m
    LEFT JOIN UserResonator ur
        ON ur.resonatorMaster = m
       AND ur.isDeleted = false
    """)
    List<ResonatorSummaryResponseDto> findResonatorSummary();

    boolean existsByName(String name);
}

File: fastapi/app/repositories/resonator_master_repository.py
from typing import List, Optional
from sqlalchemy import select, exists
from sqlalchemy.orm import Session
from app.models.resonator_master import ResonatorMaster
from app.models.user_resonator import UserResonator
from app.schemas.response import ResonatorSummaryResponseDto


def find_by_name(db: Session, name: str) -> Optional[ResonatorMaster]:
    return db.scalar(select(ResonatorMaster).where(ResonatorMaster.name == name))


def exists_by_name(db: Session, name: str) -> bool:
    return db.scalar(
        select(exists().where(ResonatorMaster.name == name))
    )


def find_resonator_summary(db: Session) -> List[ResonatorSummaryResponseDto]:
    # relationship 대신 명시적 JOIN 사용 (ResonatorMaster.user_resonators는 미사용 필드라 안 걸어둠)
    stmt = (
        select(
            UserResonator.id,
            ResonatorMaster.name,
            ResonatorMaster.rarity,
            ResonatorMaster.release_version,
            ResonatorMaster.thumbnail_image,
        )
        .select_from(ResonatorMaster)
        .join(
            UserResonator,
            (UserResonator.resonator_master_id == ResonatorMaster.id)
            & (UserResonator.is_deleted.is_(False)),
            isouter=True,
        )
    )
    rows = db.execute(stmt).all()

    return [
        ResonatorSummaryResponseDto(
            userResonatorId=row.id,
            resonatorName=row.name,
            rarity=row.rarity,
            releaseVersion=row.release_version,
            thumbnailImageUrl=row.thumbnail_image,
        )
        for row in rows
    ]
